lib/src/http/utils.dart

```diff
@@ -14,14 +14,64 @@
 
 import 'dart:async';
 import 'dart:convert';
+import 'dart:math';
 import 'dart:typed_data';
 
 import 'package:http_parser/http_parser.dart';
 
+import 'package:w_transport/src/http/auto_retry.dart';
 import 'package:w_transport/src/http/request_progress.dart';
 
 /// RegExp that only matches strings containing only ASCII-compatible chars.
-final RegExp _asciiOnly = new RegExp(r'^[\x00-\x7F]+$');
+final _asciiOnly = new RegExp(r'^[\x00-\x7F]+$');
+
+/// Based used when calculating the exponential backoff.
+const _exponentialBase = 2;
+
+/// Calculate the backoff duration based on [RequestAutoRetry] configuration.
+/// Returns [null] if backoff is not applicable.
+Duration calculateBackOff(RequestAutoRetry autoRetry) {
+  Duration backOff;
+  switch (autoRetry.backOff.method) {
+    case RetryBackOffMethod.exponential:
+      backOff = calculateExponentialBackOff(autoRetry);
+      break;
+    case RetryBackOffMethod.fixed:
+      backOff = calculateFixedBackOff(autoRetry);
+      break;
+    case RetryBackOffMethod.none:
+    default:
+      break;
+  }
+  return backOff;
+}
+
+Duration calculateExponentialBackOff(RequestAutoRetry autoRetry) {
```

Comment:
these 2 calc methods could be private instead, but since I assume this isn't actually publicly exposed (shrug)